Uploading spreadsheet for updates

aaclan-ebi opened the pull request:
ebi-ait/dcp-ingest-central#349

Commit:
Uploading spreadsheet for updates, failing tests

Commit:
Passing of submission url

Commit:
Renamed method in the importer

Commit:
Optimise imports

Commit:
Reverted requirements.txt upgrades

Commit:
Fix tests

Commit:
Updated tests

amnonkhen commented on broker/service/spreadsheet_storage/spreadsheet_storage_service.py:
```diff
@@ -18,20 +18,29 @@ def store(self, submission_uuid, spreadsheet_name, spreadsheet_blob):
 
-    def retrieve(self, submission_uuid):
+    def store_binary_file(self, directory: str, filename:str, blob):
+        path = f'{directory}/{filename}'
```
Do you need to create the output directory?

amnonkhen commented on broker/service/spreadsheet_upload_service.py:
```diff
@@ -16,28 +18,47 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
         self.importer = importer
 
-    def async_upload(self, token, request_file, is_update, project_uuid=None):
+    def async_upload(self, token, request_file, is_update, project_uuid=None, submission_uuid=None):
```
There are a few branches in this function. It is confusing to keep track of.

aaclan-ebi replied:
Refactored. Hopefully, it's more readable now!

Commit:
Make sure directory exists before saving the file

Commit:
Use makedirs not mkdir

amnonkhen commented on test/service/test_spreadsheet_generator.py:
```diff
@@ -195,39 +202,53 @@ def test_generate(self):
         actual_col_names = list(df.iloc[2])
-        self.assertEqual(expected_col_names, actual_col_names)
\ No newline at end of file
+        self.assertEqual(expected_col_names, actual_col_names)
```
Have you considered creating a "golden" spreadsheet against which to compare? In stead of having low-level comparisons?

aaclan-ebi replied:
I just added an extra line here, but i didn't modify the existing tests.

amnonkhen commented on test/service/test_spreadsheet_generator.py:
```diff
@@ -195,39 +202,53 @@ def test_generate(self):
+
+    def tearDown(self) -> None:
+        if self.test_file:
+            if os.path.exists(self.test_file):
```
I think you could use the tempfile library for tempfiles that automatically delete

aaclan-ebi replied:
cool, very nice lib! thanks!

amnonkhen commented on test/test_broker_app.py:
```diff
@@ -49,7 +52,7 @@ def test_upload_success(self, mock_ingest, mock_async_upload, mock_request):
         self.assertEqual(response.status_code, 201)
         self.assertRegex(str(response.data), 'url/9')
-        mock_async_upload.assert_called_with('token', 'content'.encode(), False, None)
+        mock_async_upload.assert_called_with('token', 'content'.encode(), False, None, None)
```
This line seems to repeat in several tests, I would suggest refactoring to a shared function.

aaclan-ebi replied:
I decided to remove them, I think these assertions are brittle and too tied with implementation. Asserting the correct HTTP code and message match should be good enough.

amnonkhen commented on test/service/test_spreadsheet_upload_service.py:
```diff
@@ -20,38 +20,24 @@ def setUp(self) -> None:
     def test_upload_success(self):
         # given:
-        self.importer.create_update_spreadsheet = Mock()
+        self.importer.update_spreadsheet_with_uuids = Mock()
```
Some of the setup code is copied in a few tests, I would refactor and create a reusable function.

aaclan-ebi replied:
I've moved them all up to the setUp function.

Commit:
Use tempfile for deleting files in the tests

Commit:
Move redundant code to setUp function

Commit:
Refactored async_upload for readability

Commit:
Remove brittle assertions

Commit:
Better function name

Commit:
Some more fixes to tests

ke4 reviewed (dismissed):
LGTM

amnonkhen commented on broker/service/spreadsheet_upload_service.py:
```diff
@@ -16,28 +18,55 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
+            raise SpreadsheetUploadError(401, "An authentication token must be supplied when uploading a spreadsheet")
+        self.ingest_api.set_token(token)
+
+    def upload(self, submission_url, path, project_uuid=None):
```
Is `upload` for a full project rather than just updates?

aaclan-ebi replied:
The upload function is for uploading spreadsheets with new entities. Sometime's the project is created before submission so we supply it if it already exists.

amnonkhen commented on broker/service/spreadsheet_upload_service.py:
```diff
@@ -16,28 +18,55 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
+            submission_resource = self.ingest_api.create_submission()
+        return submission_resource
+
+    def _assert_token_exists(self, token):
```
Is this input validation code? `assert` makes it sound like testing code.

aaclan-ebi replied:
Renamed to _validate_token_exists

amnonkhen commented on broker_app.py:
```diff
@@ -46,9 +46,9 @@
-ingest_api = IngestApi()
-spreadsheet_generator = SpreadsheetGenerator(ingest_api)
-spreadsheet_job_manager = SpreadsheetJobManager(spreadsheet_generator, SPREADSHEET_STORAGE_DIR)
+global ingest_api
```
Is using `global` a safe approach?
Should we use flask's app context instead? or the "g" object?

aaclan-ebi replied:
Ah, i didn't know about that. Applied.

aaclan-ebi replied:
Hmm it doesn't function as a global variable.

> This is a good place to store resources during a request. During testing, you can use the Faking Resources and Context pattern to pre-configure such resources.

I'll revert to using the global :(

ke4 replied:
It might be my lack of knowledge, but why are we defining these 3 variables here and once again at the `setup` method?

amnonkhen commented on test/integration/test_spreadsheet_generator.py:
```diff
@@ -12,7 +13,13 @@
 import yaml
 
+
+# TODO these are integration tests, could either move to a separate directory
```
good idea.

aaclan-ebi replied:
I removed the comments, I already moved them now to a separate directory to make it easier to run tests locally 😅

Commit:
Use latest version of ingest-client

Commit:
Rename function for clarity

Commit:
Remove comments as they're already applied

Commit:
Reverted use of global declaration and setup

Commit:
Use flask g application context instead of global variables

Commit:
Revert "Reverted use of global declaration and setup"

This reverts commit e7b8bda02ef0588ad02924d6461f497f703df950.

Commit:
Remove g usage in tests

Commit:
Remove g usage, revert commit was insufficient

Commit:
Added __init__.py to run nosetests

Commit:
Remove multiprocessing as it has a very short time out

Commit:
Use latest ingest-client version - fixed bug

ke4 reviewed (commented):
I made a tiny comment. Could we clarify it, please?
Tx

Commit:
Move setup to top

aaclan-ebi commented:
@ke4 i think you're right, they're not needed. I removed them and moved the setup function on top to make it clearer there are global vars to be initialised. I don't know how else to avoid using global vars at the moment. But, I can see the performance benefit of it that it only initialises the schema template obj which is being used in template spreadsheet generation only once instead of initialising it every request so I'd like to retain the use of global vars for now.

amnonkhen commented on broker_app.py:
```diff
@@ -239,10 +251,11 @@ def _upload_spreadsheet(is_update=False):
     try:
         logger.info('Uploading spreadsheet!')
-        submission_resource = spreadsheet_upload_svc.async_upload(token, request_file, is_update, project_uuid)
+        submission_resource = spreadsheet_upload_svc.async_upload(token, request_file, is_update, project_uuid, submission_uuid)
```
The is_update param is just passed through here, and there is subsequent branching logic further down the execution sequence.
This smells of unclean design. Maybe there would be room here for the "strategy" pattern.

aaclan-ebi replied:
I'm afraid I don't have time to implement that for this sprint.

amnonkhen commented on requirements.txt:
```diff
@@ -8,7 +8,7 @@ itsdangerous==0.24
 MarkupSafe==1.1
 werkzeug>=0.15.3
--e git+https://github.com/ebi-ait/ingest-client.git@7f04ab9a#egg=hca_ingest
+-e git+https://github.com/ebi-ait/ingest-client.git@7038dcb#egg=hca_ingest
```
Why are we not using a normal version number? (x.y.z)

aaclan-ebi replied:
We could.

aaclan-ebi replied:
This was done this way in the early development of the ingest-client python lib. It wasn't changed yet. We should also set up and release build for the ingest-client repo so that it'll release the module to PyPI every merges to master.

aaclan-ebi replied:
I filed a ticket for the ingest-client CI/CD setup: https://github.com/ebi-ait/dcp-ingest-central/issues/385

amnonkhen commented on test/integration/test_spreadsheet_generator.py:
```diff
@@ -12,7 +13,11 @@
 import yaml
 
+
 class TestSpreadsheetGenerator(TestCase):
```
This test file is very big. Can it be split to smaller logical units?

aaclan-ebi replied:
I believe we could do that later. This test suite is not part of the user story. I just tried to fix them because they're failing.

aaclan-ebi commented:
Hi @amnonkhen , I would like to merge this today so that I can deploy and hand over to a wrangler for testing, we could apply those comments later.

amnonkhen commented on broker_app.py:
```diff
@@ -46,9 +48,17 @@
-spreadsheet_generator = SpreadsheetGenerator(ingest_api)
-spreadsheet_job_manager = SpreadsheetJobManager(spreadsheet_generator, SPREADSHEET_STORAGE_DIR)
+
+def setup():
```
This is a poor person's application factory.

amnonkhen replied:
I am taking care of it on my branch.

amnonkhen commented on broker_app.py:
```diff
@@ -46,9 +48,17 @@
+def setup():
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+
+    global ingest_api
```
We should not be using globals, but sharing the instance properly between the clients.

amnonkhen replied:
I am taking care of it on my branch.

aaclan-ebi commented:
Closing in favour of PR https://github.com/ebi-ait/ingest-broker/pull/16 . This branch was already merged with that branch.
